internal/config/main/parse.go

```diff
@@ -0,0 +1,244 @@
+
+}
+
+func constructFlagDescFromStruct(typeSpec *types.Type) ([]flagDesc, error) {
+	var flagDescs []flagDesc
+	for _, member := range types.FlattenMembers(typeSpec.Members) {
+
+		extractedTags := types.ExtractCommentTags("+", member.CommentLines)
+		if hasViperExcludeTag(extractedTags) {
+			continue
+		}
+
+		desc := flagDesc{
+			Name:        getName(member),
+			Description: getDescription(member),
+			AddFlag:     true,
+		}
+
+		relevantTyp := member.Type
+
+		switch kind := relevantTyp.Kind; kind {
+		case types.Builtin:
+			switch relevantTyp {
+			case types.String:
+				desc.Type = stringType
+			case types.Int:
+				desc.Type = integerType
+			case types.Float32, types.Float64:
+				desc.Type = floatType
+			case types.Bool:
+				desc.Type = booleanType
+			default:
+				return nil, errors.Errorf("currently unsupported type %v", member.Type)
+			}
+		case types.Struct:
+			desc.Type = objectType
+			desc.AddFlag = false
+			subDesc, err := constructFlagDescFromStruct(member.Type)
+			// Template uses Parent name in mapstructure in order to
+			// correctly map config values
+			for i := range subDesc {
+				subDesc[i].Parent = desc.Name
+			}
+
+			if err != nil {
+				return nil, errors.Wrapf(err, "handling field %v", member.Name)
+			}
+			flagDescs = append(flagDescs, subDesc...)
+		case types.Slice:
+			desc.Type = stringSliceType
+			subDesc, err := constructFlagDescFromStruct(member.Type)
+			if err != nil {
+				return nil, errors.Wrapf(err, "handling field %v", member.Name)
+			}
+			flagDescs = append(flagDescs, subDesc...)
+		default:
+			return nil, errors.Errorf("currently unsupported type %v", member.Type)
+		}
+
+		flagDescs = append(flagDescs, desc)
+	}
+	return flagDescs, nil
+}
+
+func mainCmd() error {
+	b := parser.New()
+
+	// This avoids parsing generated files in the package (since we add +build !flagcodegen to them,
+	// which makes the parsing much quicker since the parser doesn't have to load any imported packages).
+	b.AddBuildTags("flagcodegen")
+	if err := b.AddDir("./internal/config"); err != nil {
+		return err
+	}
+	typeUniverse, err := b.FindTypes()
+	if err != nil {
+		return err
+	}
+
+	pkgNames := b.FindPackages()
+	if len(pkgNames) != 1 {
+		return errors.Errorf("found unexpected number of packages in %+v: %d", pkgNames, len(pkgNames))
+	}
+
+	pkg := typeUniverse.Package(pkgNames[0])
+	paramsType := pkg.Type(paramsStructName)
+
+	flagDescs, err := constructFlagDescFromStruct(paramsType)
+	if err != nil {
+		return err
+	}
+
+	var templateObj []templateElem
+
+	for _, flagDesc := range flagDescs {
+		buf := bytes.NewBuffer(nil)
+		enc := json.NewEncoder(buf)
+		enc.SetIndent("", "\t")
+		if err := enc.Encode(flagDesc); err != nil {
+			return errors.Wrapf(err, "couldn't marshal param %v", flagDesc)
+		}
+
+		templateObj = append(templateObj, templateElem{
+			FlagDesc:  flagDesc,
+			ParamJSON: buf.String(),
+		})
+	}
+
+	outFileName := filepath.Join(".", "internal", "config", "flags.go")
```

Comment:
When you add a go generate directive, you will need to make this path relative to the path of the file containing the go generate directive.